[WFLY-17451] add filter/filter-ref capabilites to enforce runtime req…

baranowb opened the pull request:
Issue: https://issues.redhat.com/browse/WFLY-17451

baranowb committed:
[WFLY-17451] add filter/filter-ref capabilites to enforce runtime requirements and avoid breaking configuration

bstansberry commented on undertow/src/main/java/org/wildfly/extension/undertow/filters/FilterAdd.java:
```diff
@@ -54,12 +52,11 @@ class FilterAdd extends AbstractAddStepHandler {
-        sb.setInitialMode(ServiceController.Mode.ON_DEMAND);
-        sb.install();
+        csb.setInstance(new FilterProviderService(serviceConsumer, wrapper));
+        csb.setInitialMode(ServiceController.Mode.ACTIVE);
```
I don't think changing the mode should be necessary for this JIRA. ON_DEMAND is the natural fit for the use of this service.

baranowb replied:
Well, idea was to have it more pro-active about error reporting, but I dont have strong inclination to die on this hill

bstansberry replied:
FilterProviderService doesn't really do anything in start and has no dependencies, so it's not going to error.

pferraro commented on undertow/src/main/java/org/wildfly/extension/undertow/filters/FilterAdd.java:
```diff
@@ -54,12 +53,20 @@ class FilterAdd extends AbstractAddStepHandler {
-        sb.setInstance(Service.newInstance(serviceConsumer, wrapper));
-        sb.setInitialMode(ServiceController.Mode.ON_DEMAND);
-        sb.install();
+        csb.setInstance(new FilterProviderService(serviceConsumer, wrapper));
```
Why introduce a redundant FilterProviderService class?  The implementation is identical to the previous logic.

pferraro commented on undertow/src/main/java/org/wildfly/extension/undertow/filters/FilterAdd.java:
```diff
@@ -54,12 +53,20 @@ class FilterAdd extends AbstractAddStepHandler {
+        csb.install();
     }
+
+    @Override
+    protected void recordCapabilitiesAndRequirements(OperationContext context, ModelNode operation, Resource resource)
+            throws OperationFailedException {
+        super.recordCapabilitiesAndRequirements(context, operation, resource);
+        //NOTE: its dynamic by design.
+        context.registerCapability(FilterCapabilities.FILTER_CAPABILITY.getDefinition().fromBaseCapability(context.getCurrentAddress()));
+    }
+
```
You do not need to override this method.  Capability registration is already handled via the super implementation.
See: https://github.com/wildfly/wildfly-core/blob/main/controller/src/main/java/org/jboss/as/controller/AbstractAddStepHandler.java#L216

baranowb replied:
AFAIR Brian said it has to be done this way.

pferraro replied:
Provided that the filter's ResourceDefinition registers the RuntimeCapability (and it should), the capability will be auto-registered with the OperationContext by the super implementation.

bstansberry replied:
@baranowb @pferraro My Sept 7, 2023 comment about using this method was in reference to FilterRefAdd, not FilterAdd.

The issue is this:

https://github.com/wildfly/wildfly/pull/17134/files#diff-8a27bf46cb96032958c3cb8549c72bb98a745b73e2f3aa2756eaeb698f532a1fR130

```
final Supplier<HandlerWrapper> fSupplier = csb.requiresCapability(Capabilities.CAPABILITY_FILTER, HandlerWrapper.class, name);
```

This is a capability requirement that is not enforced at Stage.MODEL; it's only enforced at Stage.RUNTIME via this service dependency. So an admin-only server can be used to create an invalid config that's only detected when the config is switched out of admin-only. And the bad config doesn't abort boot, which it should.

The issue is this capability reference is not done via an attribute; it's the filter-ref resource's own name that specifies the name of the required filter. So it's at the level of the filter-ref resource that we need to record the capability requirement in Stage.MODEL. AbstractAddStepHandler. recordCapabilitiesAndRequirements is the place I could think of where such a requirement can be registered.

pferraro commented on undertow/src/main/java/org/wildfly/extension/undertow/filters/FilterCapabilities.java:
```diff
@@ -0,0 +1,90 @@
+enum FilterCapabilities implements Capability {
+
+    FILTER_CAPABILITY(CAPABILITY_FILTER, HandlerWrapper.class),
+    FILTER_REF_CAPABILITY(CAPABILITY_FILTER_REF, UndertowFilter.class,new Function<PathAddress, String[]>() {
+
+        @Override
+        public String[] apply(PathAddress t) {
+            // /subsystem=undertow/server=default-server/host=fridge-server/filter-ref=XXX
+            // org.wildfly.extension.undertow.filter-ref.default-server.fridge-server.XXXX
+            // /subsystem=undertow/server=default-server/host=fridge-server/location=test/filter-ref=XXX
+            // org.wildfly.extension.undertow.filter-ref.default-server.fridge-server.test.XXXX
+
+            final List<String> elements = new ArrayList<String>();
+            boolean include = false;
+            final Iterator<PathElement> it = t.iterator();
+            while(it.hasNext()) {
+                PathElement e = it.next();
+                if(include) {
+                    elements.add(e.getValue());
+                } else if(e.getKey().equals(ModelDescriptionConstants.SERVER)){
+                    elements.add(e.getValue());
+                    include = true;
+                }
+            }
+            return elements.toArray(new String[elements.size()]);
+        }
+    });
```
I do not think you can reuse the same RuntimeCapability implementation for both contexts (i.e. per-host vs per-location), since the name resolution logic if different, as is the number of dynamic segments.  This will otherwise break galleon.  I ran into a similar problem when intitially modeling capabilities for default caches in the Infinispan subsystem.  Essentially, a capability with a given name must have a fixed number of name segments.

Rather, you'll want to parameterize the FilterRefDefinition with its RuntimeCapability.  That way the host can register the variant with 3 dynamic name segments, and the location can register the variant with 4 dynamic name segments.

FYI, once https://github.com/wildfly/wildfly-core/pull/5516 is merged, you can use [1] for the host filter-ref and [2] for the location filter-ref.

[1] https://github.com/pferraro/wildfly-core/blob/WFCORE-6311/controller/src/main/java/org/jboss/as/controller/capability/TernaryCapabilityNameResolver.java#L27
[2] https://github.com/pferraro/wildfly-core/blob/WFCORE-6311/controller/src/main/java/org/jboss/as/controller/capability/QuaternaryCapabilityNameResolver.java#L28

baranowb replied:
Ok, I will rebase it once #5516 is in.

pferraro commented on undertow/src/main/java/org/wildfly/extension/undertow/filters/FilterRefCapability.java:
```diff
@@ -0,0 +1,50 @@
+import org.jboss.as.controller.capability.RuntimeCapability;
+import org.jboss.as.controller.descriptions.ModelDescriptionConstants;
+
+class FilterRefCapability implements Capability {
```
What is the purpose of this class?  It does not seem to be referenced anywhere.

baranowb replied:
ha. power out, thought it did not save and must have picked another name.

pferraro commented on undertow/src/main/java/org/wildfly/extension/undertow/filters/FilterRefDefinition.java:
```diff
@@ -128,15 +131,44 @@ protected void performRuntime(OperationContext context, ModelNode operation, Mod
             sb.install();
         }
+
+        @Override
+        protected void recordCapabilitiesAndRequirements(OperationContext context, ModelNode operation, Resource resource)
+                throws OperationFailedException {
+            super.recordCapabilitiesAndRequirements(context, operation, resource);
+            //NOTE: its dynamic by design.
+            context.registerCapability(FilterCapabilities.FILTER_REF_CAPABILITY.getDefinition().fromBaseCapability(context.getCurrentAddress()));
+        }
```
You do not need to override this method. Capability registration is already handled via the super implementation.
See: https://github.com/wildfly/wildfly-core/blob/main/controller/src/main/java/org/jboss/as/controller/AbstractAddStepHandler.java#L216

pferraro commented on undertow/src/main/java/org/wildfly/extension/undertow/filters/FilterRemove.java:
```diff
@@ -0,0 +1,53 @@
+/**
+ * @author baranowb
+ */
+public class FilterRemove extends ServiceRemoveStepHandler {
+
+    public FilterRemove(AbstractAddStepHandler addOperation) {
+        super(addOperation);
+    }
+
+    public FilterRemove(ServiceName baseServiceName, AbstractAddStepHandler addOperation) {
+        super(baseServiceName, addOperation);
+    }
+
+    @Override
+    protected void recordCapabilitiesAndRequirements(OperationContext context, ModelNode operation, Resource resource)
+            throws OperationFailedException {
+        super.recordCapabilitiesAndRequirements(context, operation, resource);
+
+        context.deregisterCapability(FilterCapabilities.FILTER_CAPABILITY.getDefinition().getDynamicName(context.getCurrentAddress()));
+    }
+
+}
```
You do not need to subclass ServiceRemoveStepHandler.
Capability deregistration is already handled via the super implementation.
See: https://github.com/wildfly/wildfly-core/blob/main/controller/src/main/java/org/jboss/as/controller/AbstractRemoveStepHandler.java#L175

baranowb replied:
Same with this one, explicitly requested by Brian.

pferraro replied:
Provided that the filter's ResourceDefinition registered the RuntimeCapability (and it should), the capability will be auto-deregistered with the OperationContext by the super implementation.

pferraro commented on undertow/src/main/java/org/wildfly/extension/undertow/filters/FilterRefDefinition.java:
```diff
@@ -128,15 +131,44 @@ protected void performRuntime(OperationContext context, ModelNode operation, Mod
+            context.registerCapability(FilterCapabilities.FILTER_REF_CAPABILITY.getDefinition().fromBaseCapability(context.getCurrentAddress()));
+        }
+
+    }
+
+    static class FilterRefRemove extends ServiceRemoveStepHandler{
+
+        public FilterRefRemove(ServiceName baseServiceName, AbstractAddStepHandler addOperation) {
+            super(baseServiceName, addOperation);
+        }
+
+        public FilterRefRemove(AbstractAddStepHandler addOperation) {
+            super(addOperation);
+        }
+
+        @Override
+        protected void recordCapabilitiesAndRequirements(OperationContext context, ModelNode operation, Resource resource)
+                throws OperationFailedException {
+            super.recordCapabilitiesAndRequirements(context, operation, resource);
+            context.deregisterCapability(FilterCapabilities.FILTER_REF_CAPABILITY.getDefinition().getDynamicName(context.getCurrentAddress()));
+        }
+
```
You do not need to subclass ServiceRemoveStepHandler.
Capability deregistration is already handled via the super implementation.
See: https://github.com/wildfly/wildfly-core/blob/main/controller/src/main/java/org/jboss/as/controller/AbstractRemoveStepHandler.java#L175

baranowb commented:
So far Im unable to reproduce this failure manually.

bstansberry commented:
@baranowb @fl4via @pferraro My impression is this one could have gotten in but just stalled out. Can we get it done?

I'd like to get the PR queue down to around 40 open PRs, about 1/2 the current size.

baranowb commented:
ATM Im stuck at: "WFLYCTL0158: Operation handler failed: java.lang.IllegalArgumentException: WFLYCTL0394 - either I missed something or it class with subsequent changes.

bstansberry committed:
[WFLY-17451] Register a requirement for the filter capability in the Stage.MODEL handling for adding a filter-ref.

Unfortunately, Galleon feature spec generation won't let us record it via FilterResourceRegistration.

Also, statically parameterize FilterRefDefinition and FilterRefAdd instances so they know if they are for direct children of 'host' or for children of 'location'

bstansberry reviewed (changes requested):
Thanks @baranowb. Besides the header issue I noted, I sent you https://github.com/baranowb/wildfly/pull/1 to illustrate how the filter-ref requirement for the filter needs to be done. With what's here now the system will not detect that the config is wrong until it tries to do service wiring. So it won't detect it in an --admin-only server, where that service wiring code is not executed.

The way this needs to be handled is not ideal because of how Galleon works, but what I did is next best. Comments on my PR to you explain what I mean.

baranowb commented:
Thanks. I will compare to what I have locally, since it is still failing, I might be missing something that you have set up properly.

baranowb commented on undertow/src/main/java/org/wildfly/extension/undertow/filters/FilterCapabilities.java:
```diff
@@ -0,0 +1,64 @@
+ * @author Paul Ferraro
+ * @author baranowb
+ */
+enum FilterCapabilities implements Capability {
```
https://github.com/wildfly/wildfly/blob/main/clustering/common/src/main/java/org/jboss/as/clustering/controller/Capability.java  ? will be ?

bstansberry commented:
/retest

bstansberry commented:
Thanks @baranowb -- I'm closing this because #20249 picked it up and is merged. Sorry for forgetting to come back to this.
